godot-core/src/builtin/dictionary.rs

```diff
@@ -0,0 +1,350 @@
+    /// no value with the given key.
+    pub fn get<K: ToVariant>(&self, key: K) -> Option<Variant> {
+        let key = key.to_variant();
+        if !self.contains_key(key.clone()) {
+            return None;
+        }
+
+        Some(self.get_or_nil(key))
+    }
+
+    /// Returns the value at the key in the dictionary, or nil otherwise. This
+    /// method does not let you differentiate `NIL` values stored as values from
+    /// absent keys. If you need that, use `get()`.
+    ///
+    /// This is equivalent to `get` in Godot.
+    pub fn get_or_nil<K: ToVariant>(&self, key: K) -> Variant {
+        self.as_inner().get(key.to_variant(), Variant::nil())
+    }
+
+    /// Returns `true` if the dictionary contains the given key.
+    ///
+    /// This is equivalent to `has` in Godot.
+    pub fn contains_key<K: ToVariant>(&self, key: K) -> bool {
+        let key = key.to_variant();
+        self.as_inner().has(key)
+    }
+
+    /// Returns `true` if the dictionary contains all the given keys.
+    ///
+    /// This is equivalent to `has_all` in Godot.
+    pub fn contains_all_keys(&self, keys: Array) -> bool {
+        self.as_inner().has_all(keys)
+    }
+
+    /// Returns a 32-bit integer hash value representing the dictionary and its contents.
+    pub fn hash(&self) -> u32 {
+        self.as_inner().hash().try_into().unwrap()
+    }
+
+    /// Creates a new `Array` containing all the keys currently in the dictionary.
+    pub fn keys(&self) -> Array {
+        self.as_inner().keys()
+    }
+
+    /// Creates a new `Array` containing all the values currently in the dictionary.
+    pub fn values(&self) -> Array {
+        self.as_inner().values()
+    }
+
+    /// Returns true if the dictionary is empty.
+    pub fn is_empty(&self) -> bool {
+        self.as_inner().is_empty()
+    }
+
+    /// Copies all keys and values from other into self.
+    ///
+    /// If overwrite is true, it will overwrite pre-existing keys. Otherwise
+    /// it will not.
+    ///
+    /// This is equivalent to `merge` in Godot.
+    pub fn extend_dictionary(&mut self, other: Self, overwrite: bool) {
+        self.as_inner().merge(other, overwrite)
+    }
+
+    /// Returns the number of entries in the dictionary.
+    ///
+    /// This is equivalent to `size` in Godot.
+    pub fn len(&self) -> usize {
+        self.as_inner().size().try_into().unwrap()
+    }
+
+    /// Get the pointer corresponding to the given key in the dictionary,
+    /// this pointer is null if there is no value at the given key.
+    #[allow(dead_code)] // TODO: remove function if it turns out i'll never actually get any use out of it
+    fn get_ptr<K: ToVariant>(&self, key: K) -> *const Variant {
+        let key = key.to_variant();
+        unsafe {
+            let ptr = (interface_fn!(dictionary_operator_index_const))(
+                self.sys_const(),
+                key.var_sys_const(),
+            );
+            ptr as *const Variant
+        }
```

Comment:
In contrast to `get_ptr_mut`, this function has no `assert!` statement for the nullability of the pointer; maybe add it to not cause UB down the line.

Reply:
but this pointer can be null, and that indicates that there is no key with the associated value. `get_ptr_mut` will actually never be null however.